### PR TITLE
Functional Tests - Add BO tests for generate invoice by date

### PR DESCRIPTION
**Pull request by nesrineabdmouleh**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Add Functional tests for generate invoice by date
| Type?         | refacto
| Category?     | TE
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | no
| How to test?  | TEST_PATH="functional/BO/orders/invoices/01_generateInvoiceByDate.js" URL_FO=yourShopURL DOWNLOAD_PATH=yourDownloadPath npm run specific-test

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/16566)

### CODE REVIEW

File: tests/puppeteer/.eslintrc.js [hunk 1]
```diff
@@ -16,10 +16,11 @@ module.exports = {
...
   },
   rules: {
-    'no-plusplus': [2, { allowForLoopAfterthoughts: true }],
+    'no-plusplus': [2, {allowForLoopAfterthoughts: true}],
     'func-names': 'off',
     'no-await-in-loop': 'off',
     'class-methods-use-this': ['error', {'exceptMethods': ['replaceAll']}],
-    'max-len': [2, {code: 120}]
+    'max-len': [2, {code: 120}],
+    'no-underscore-dangle': 'off'
```
Comment (boubkerbribri) [inline tests/puppeteer/.eslintrc.js:24]: Is this linked to this ?
```js
await browser._client.send('Page.setDownloadBehavior', {
      behavior: 'allow',
      downloadPath: global.BO.DOWNLOADSPATH,
    });
```

File: tests/puppeteer/pages/BO/orders/invoices/index.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    this.dateFromInput = `${this.generateByDateForm} #form_generate_by_date_date_from`;
+    this.dateToInput = `${this.generateByDateForm} #form_generate_by_date_date_to`;
+    this.generatePdfByDateButton = `${this.generateByDateForm} .btn.btn-primary`;
+  }
+
+  /*
+  Methods
+   */
+
+  async generatePDFByDate(dateFrom = '', dateTo = '') {
```
Comment (boubkerbribri) [inline tests/puppeteer/pages/BO/orders/invoices/index.js:28]: please add function description

File: tests/puppeteer/pages/commonPage.js [hunk 3]
```diff
@@ -253,4 +256,21 @@ module.exports = class CommonPage {
...
+   */
+  async checkFileExistence(downloadPath, fileName, timeDelay = 5000) {
+    for (let i = 0; i < timeDelay; i++) {
+      await this.page.waitFor(1);
+      if (fs.existsSync(`${downloadPath}/${fileName}`)) {
+        break;
+      }
+    }
+    return fs.existsSync(`${downloadPath}/${fileName}`);
+  }
```
Comment (boubkerbribri) [inline tests/puppeteer/pages/commonPage.js]:
```js
  async checkFileExistence(downloadPath, fileName, timeDelay = 5000) {
    let found = false;
    for (let i = 0; i < timeDelay && !found; i += 10) {
      setTimeout(() => { /* only to sleep */ }, 10);
      found = await this.isFileExist(downloadPath, fileName);
    }
    return found;
  }
```
Maybe you should :
- reduce frequency of testing if file exist (to 10 or 20 ms not 1)
- use utils/files.js and not the common page
- use var found to not check the existence of the file again while returning value.
(to be tested)

File: tests/puppeteer/pages/BO/BObasePage.js [hunk 4]
```diff
@@ -76,6 +78,7 @@ module.exports = class BOBasePage extends CommonPage {
     // Alert Text
     this.alertSuccessBloc = 'div.alert.alert-success:not([style=\'display: none;\'])';
     this.alertSuccessBlockParagraph = `${this.alertSuccessBloc} div.alert-text p`;
+    this.alertTextBlock = '.alert-text';
```
Comment (boubkerbribri) [inline tests/puppeteer/pages/BO/BObasePage.js:81]: it's some thing different than `alertSuccessBlockParagraph` ?

File: tests/puppeteer/campaigns/utils/helpers.js [hunk 5]
```diff
@@ -12,4 +12,10 @@ module.exports = {
   async closeBrowser(browser) {
     return browser.close();
   },
+  async setDownloadBehavior(browser) {
+    await browser._client.send('Page.setDownloadBehavior', {
+      behavior: 'allow',
+      downloadPath: global.BO.DOWNLOADSPATH,
+    });
+  },
```
Comment (boubkerbribri) [inline tests/puppeteer/campaigns/utils/helpers.js]: :+1:

File: tests/puppeteer/pages/commonPage.js [hunk 6]
```diff
@@ -17,6 +19,7 @@ module.exports = class CommonPage {
    * @return textContent
    */
   async getTextContent(selector) {
+    await this.page.waitForSelector(selector, {visible: true});
```
Comment (boubkerbribri) [inline tests/puppeteer/pages/commonPage.js:20]: No need to add this here, I think you should wait for your selector after the click in function `generatePDFByDate`

File: tests/puppeteer/campaigns/functional/BO/orders/invoices/01_generateInvoiceByDate.js [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+          Invoices.moreThanAnInvoice.fileName,
+        );
+        await expect(exist).to.be.false;
+      });
+
+      it('should check the error message when there is no invoice in the entered date', async function () {
+        await this.pageObjects.invoicesPage.generatePDFByDate(dateFrom, dateTo);
+        const textMessage = await this.pageObjects.invoicesPage.getTextContent(
+          this.pageObjects.invoicesPage.alertTextBlock,
+        );
```
Comment (boubkerbribri) [inline tests/puppeteer/campaigns/functional/BO/orders/invoices/01_generateInvoiceByDate.js:107]: You should return message in function `generatePDFByDate`

**Review by boubkerbribri — COMMENTED** [review]
thank you @nesrineabdmouleh

File: tests/puppeteer/.eslintrc.js (hunk 1 above)
Comment (nesrineabdmouleh) [reply tests/puppeteer/.eslintrc.js:24]: yes

File: tests/puppeteer/pages/BO/BObasePage.js (hunk 4 above)
Comment (nesrineabdmouleh) [reply tests/puppeteer/pages/BO/BObasePage.js:81]: Yes it's different, it's not the success alert, I use it for the error message

File: tests/puppeteer/pages/commonPage.js (hunk 6 above)
Comment (nesrineabdmouleh) [reply tests/puppeteer/pages/commonPage.js:20]: There is no validation message all the time, the error message is displayed only when there is no invoice in the selected date.

File: tests/puppeteer/campaigns/functional/BO/orders/invoices/01_generateInvoiceByDate.js (hunk 7 above)
Comment (nesrineabdmouleh) [reply tests/puppeteer/campaigns/functional/BO/orders/invoices/01_generateInvoiceByDate.js:107]: The message is not displayed all the time.

File: tests/puppeteer/README.md [hunk 8]
```diff
@@ -24,6 +24,7 @@ npm install
 | DB_USER             | Login user of your MySql (default to **`root`**) |
 | DB_PASSWD           | Password for your MySql (default to **`empty`**) |
 | HEADLESS            | Boolean to run tests in headless or not (default to **`true`**) |
+| DOWNLOADSPATH       | PATH of your download folder (default to **`empty`**)|
```
Comment (PierreRambaud) [inline tests/puppeteer/README.md]:
```suggestion
| DOWNLOAD_PATH       | PATH of your download folder (default to **`empty`**)|
```

File: tests/puppeteer/campaigns/utils/files.js [hunk 9]
```diff
@@ -9,4 +9,14 @@ module.exports = {
...
   },
+
+  /**
+   * Check File existence
+   * @param downloadPath
+   * @param fileName
+   * @return boolean, true if exist, false if not
+   */
+  async isFileExist(downloadPath, fileName) {
+    return fs.existsSync(`${downloadPath}/${fileName}`);
```
Comment (PierreRambaud) [inline tests/puppeteer/campaigns/utils/files.js]: is it really useful to create a function to call a function :thinking: 
it's like doing this
```
function consolelog(downloadPath, fileName) {
  console.log(`${downloadPath}/${fileName}`);
}
```

File: tests/puppeteer/campaigns/utils/globals.js [hunk 10]
```diff
@@ -9,6 +9,7 @@ global.BO = {
   PASSWD: process.env.PASSWD || DefaultAccount.password,
   FIRSTNAME: process.env.FIRSTNAME || DefaultAccount.firstName,
   LASTNAME: process.env.LASTNAME || DefaultAccount.lastName,
+  DOWNLOADSPATH: process.env.DOWNLOADSPATH || '',
```
Comment (PierreRambaud) [inline tests/puppeteer/campaigns/utils/globals.js]:
```suggestion
  DOWNLOAD_PATH: process.env.DOWNLOAD_PATH || '',
```

File: tests/puppeteer/campaigns/utils/helpers.js [hunk 11]
```diff
@@ -12,4 +12,10 @@ module.exports = {
   async closeBrowser(browser) {
     return browser.close();
   },
+  async setDownloadBehavior(browser) {
+    await browser._client.send('Page.setDownloadBehavior', {
+      behavior: 'allow',
+      downloadPath: global.BO.DOWNLOADSPATH,
```
Comment (PierreRambaud) [inline tests/puppeteer/campaigns/utils/helpers.js]:
```suggestion
      downloadPath: global.BO.DOWNLOAD_PATH,
```

File: tests/puppeteer/pages/commonPage.js [hunk 12]
```diff
@@ -253,4 +256,21 @@ module.exports = class CommonPage {
...
+  /**
+   * Check File existence
+   * @param downloadPath
+   * @param fileName
+   * @param timeDelay
+   * @return boolean, true if exist, false if not
+   */
+  async checkFileExistence(downloadPath, fileName, timeDelay = 5000) {
+    for (let i = 0; i < timeDelay; i++) {
+      await this.page.waitFor(1);
```
Comment (PierreRambaud) [inline tests/puppeteer/pages/commonPage.js]: No need to wait 1ms each time.
You can wait 500ms and loop ten times instead

File: tests/puppeteer/pages/commonPage.js (hunk 6 above)
Comment (boubkerbribri) [reply tests/puppeteer/pages/commonPage.js:20]: Still, it's weird to have waitForSelector here, it will be duplicated in some functions

**Review by boubkerbribri — APPROVED** [review]
thanks @nesrineabdmouleh

**Review by PierreRambaud — APPROVED** [review]

**Comment by matks** [comment]
@boubkerbribri @SimonGrn When you merged tests PR, could you please add the related milestone to the PR 😉? This helps us to build statistics later for releases

**Pull request merged**